### DISCUSSION TITLE
GitSub Project: Sorting function only sorts objects already on the DOM, not newly added objects.

### DISCUSSION THREAD

**Discussion by Gabrielle-Tobermann** [nss-evening-cohort-14/evening-client-side#96] [Help Me!] [answered]
Name: Gabby Tobermann 

Search through other tickets: Yes, I read through a few tickets but did not find anything that would help. 

Issue: I'm working on the stretch goal for the projects page of the team project. I'm trying to sort through my projects. My sorting function works fine with the five projects which are already displayed on the DOM. But when I add a new project and try to sort again, it only sorts those 5 original projects and ignores the new ones. I am storing my newly sorted projects in a variable called 'sortedProjects'. I think my problem is, when I add a new array it pushes the object to my projects array, while my sorting function sorts my projects, puts them in my sortedProjects array and rebuilds the cards with my sortedProjects. I am not sure how to work around that. 

I do not get any errors. 

Google links: https://stackoverflow.com/questions/59789796/check-if-array-of-objects-has-changed
https://stackoverflow.com/questions/5100376/how-to-watch-for-array-changes
https://stackoverflow.com/questions/5100376/how-to-watch-for-array-changes
https://gomakethings.com/attaching-multiple-elements-to-a-single-event-listener-in-vanilla-js/

What I have tried: I've read all documentation I could find on sort method, event listeners. 
- I have tried just sorting my projects array without storing in another variable, that didn't work.
- I have tried adding my sorting functionality to my projectsFormInfo without success. 
- I searched how to check if an array has changed, thinking I could create an if statement that would look something like: 
  if (projects array has changed) {
     sort again
} 
But I could not find a way to do that. 
I'm sure I'm just missing something super simple, but I cannot figure it out! 

Here is a link to my JS file: https://github.com/nss-evening-cohort-14/gitsub-e14-team-2-electric-boogaloo/blob/gt-projects-stretch-sort/main.js

Relevant functions: 
- createCards line 262
- projectsFormInfo line 488
- sortProjectsCards line 507
- pageInit line 515
- variables are declared at the top of the JS file.

**Comment by drteresavasquez** [comment]
I'll need to look more closely, but you need to make sure that the sorted array is passed to the print to DOM function otherwise it will just place them on the DOM in order. 

Have you tried sorting the array before updating the DOM with the array?

**Reply by drteresavasquez** [reply]
Also here:

```javascript
//Gabby stretch goal - sort cards 
const sortProjectCards = (e) => {

  if (e.target.id === 'sort-btn') {
       sortedProjects = projects.slice().sort((a, b) => b.date - a.date);
  };
  createCards(sortedProjects, projectCards, '#project-container');
};
```

What is the purpose of the slice method? Also here, you have the sort happening on the sort button. Is that the ID of the button that is adding projects?

**Reply by Gabrielle-Tobermann** [reply]
I added slice because without it, I could only sort once and then the button wouldn't work after clicking it once. With slice, the cards reset when a new object is added.  I thought it was a step closer to fixing my problem, but might have been the wrong approach. Same with the button, originally I was targeting my project container but nothing was happening. Somehow when I target the ID on the button, they sort (but only the first 5 projects). I will check to see if my new array is passed to the DOM, I'm pretty sure it's not.

**Reply by Gabrielle-Tobermann** [reply]
I am calling printToDom inside of my createCards function. So when I pass my new array to createCards, they appear. When I fill out the form, a new project shows up at the bottom of my project cards. When I submit that new project, my projects return to their origin position (unsorted). When I clicked the sort button again, they sort, except for the newly added project, which stays at the bottom.

**Reply by Gabrielle-Tobermann** [reply]
Update: I found the problem. when pushing a new object to my cards, I had defined the date property like this: const date = Date(). But I actually needed date = new Date().

**Reply by drteresavasquez** [reply]
Wahooo!!!!! Great job, @Gabrielle-Tobermann! Make your answer a direct comment so that I can add it as the answer. WTG! 🥳

**Comment by Gabrielle-Tobermann** [comment, marked as answer]
Update: I found the problem. when pushing a new object to my cards, I had defined the date property like this: const date = Date(). But I actually needed date = new Date().
